model/map/src/main/java/org/keycloak/models/map/storage/chm/UserSessionConcurrentHashMapStorage.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.models.map.storage.chm;
+
+import org.keycloak.models.AuthenticatedClientSessionModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.UserSessionModel;
+import org.keycloak.models.map.common.AbstractEntity;
+import org.keycloak.models.map.storage.MapKeycloakTransaction;
+import org.keycloak.models.map.storage.ModelCriteriaBuilder;
+import org.keycloak.models.map.storage.ModelCriteriaBuilder.Operator;
+import org.keycloak.models.map.userSession.AbstractAuthenticatedClientSessionEntity;
+import org.keycloak.models.map.userSession.AbstractUserSessionEntity;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * User session storage with a naive implementation of referential integrity in client to user session relation, restricted to
+ * ON DELETE CASCADE functionality.
+ *
+ * @author hmlnarik
+ */
+public class UserSessionConcurrentHashMapStorage<K> extends ConcurrentHashMapStorage<K, AbstractUserSessionEntity<K>, UserSessionModel> {
```

Comment:
Have you considered using some listener of `UserSessionRemoved` events in `MapUserSessionProviderFactory`? Maybe we can avoid sub-storages like this if we publish events automatically as part of MapKeycloakTransaction methods. WDYT? Is there something I am missing?

Reply:
Yup. The primary aim is to avoid roundtrip to Java and use referential integrity constraints known e.g. from relational databases. Hence the referential integrity implementation here - no high-level Java code intervention is desired here.

Reply:
I see. However, I still can't see the advantage of this approach. Will there be any roundtrip to Java if there is no listener registered in MapJpaUserSessionFactory? The advantage of using events is, that it is easily extendible for other storages, when they want to react upon RemovedUserSession event.

Also, there will be some refactoring of events that are currently published from providers. We plan to publish them from underlying storage, if I remember correctly. Does it mean it will be published from MapKeycloakTransaction? If yes how we will decide when to publish event and when not to? I am asking because if we will be publishing events for each storage in the future, we may end up with what I described in the first comment anyway.

Reply:
We discussed this with @hmlnarik and we agreed this is the best solution we can have at the moment.